Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h

```diff
@@ -0,0 +1,438 @@
+            }
+            ++index;
+            return map;
+        }
+        else if constexpr (std::is_same_v<T, CLuaFunctionRef>)
+        {
+            return luaM_toref(L, index++);
+        }
+        else if constexpr (std::is_same_v<T, lua_State*>)
+            return L;
+        // variants can be used by any of the underlying types
+        // thus recursively use this function
+        else if constexpr (is_variant<T>::value)
+        {
+            int iMatch = TypeMatchVariant<T>(L, index);
+            return PopUnsafeVariant<T>(L, index, iMatch);
+        }
+
+        // Catch all for class pointer types, assume all classes are valid script entities
+        // and can be fetched from a userdata
+        else if constexpr (std::is_pointer_v<T> && std::is_class_v<std::remove_pointer_t<T>>)
+        {
+            bool  isLightUserData = lua_type(L, index) == LUA_TLIGHTUSERDATA;
+            void* pValue = lua::PopTrivial<void*>(L, index);
+            using class_t = std::remove_pointer_t<T>;
+            auto result =
+                isLightUserData ? UserDataCast<class_t>((class_t*)0, pValue, L) : UserDataCast<class_t>((class_t*)0, *reinterpret_cast<void**>(pValue), L);
+            if (result == nullptr)
+            {
+                SString strReceived = isLightUserData ? GetUserDataClassName(pValue, L) : GetUserDataClassName(*(void**)pValue, L);
+                SString strExpected = GetClassTypeName((T)0);
+                SString strMessage("Expected %s at argument %d, got %s", strExpected.c_str(), index, strReceived.c_str());
+                strError = strMessage;
+                return nullptr;
+            }
+            return static_cast<T>(result);
+        }
+    }
+};
+
+template <bool, auto*>
+struct CLuaFunctionParser
+{
+};
+
+template <bool ErrorOnFailure, typename Ret, typename... Args, auto (*Func)(Args...)->Ret>
+struct CLuaFunctionParser<ErrorOnFailure, Func> : CLuaFunctionParserBase
+{
+    template <typename... Params>
+    inline auto Call(lua_State* L, Params&&... ps)
+    {
+        if (strError.length() != 0)
+        {
+            return -1;
+        }
+        if constexpr (sizeof...(Params) == sizeof...(Args))
+        {
+            if constexpr (std::is_same_v<Ret, void>)
+            {
+                Func(std::forward<Params>(ps)...);
+                return 0;
+            }
+            else
+            {
+                return lua::Push(L, Func(std::forward<Params>(ps)...));
+            }
+        }
+        else
+        {
+            return Call(L, ps..., Pop<typename nth_element_impl<sizeof...(Params), Args...>::type>(L, iIndex));
+        }
+    }
+
+    inline int operator()(lua_State* L, CScriptDebugging* pScriptDebugging)
+    {
+        int iResult = 0;
+        try
+        {
+            iResult = Call(L);
+        }
+        catch (std::invalid_argument& e)
+        {
+            strError = e.what();
```

Comment:
What happens if `strError` already has an error message? Is that even possible? If so, can we somehow emit both error messages (if useful)?

Reply:
The invalid argument exception can only be thrown from the called function (e.g. invalid hash type parameter for `passwordHash`), thus at that point `strError` is empty, as all arguments presented were valid

Reply:
Consider presenting that tautology in the form of a code comment:
```suggestion
            // This exception can only be thrown from the called function
            // (e.g. invalid hash type parameter for passwordHash)
            // therefore, as all arguments presented were valid, we can
            // guarantee `strError` is currently unset and is safe to overwrite 
            strError = e.what();
```